Add support for Sift Verification Product

dereksweet opened the pull request:
#### What does this PR do & why?

This PR adds support for the Sift verification product, their 2FA offering that will deliver one time passwords to customers for security reasons. Is intended to be used for login but can be applied to any event they support. 

#### How can this be manually tested?

You will need to become quite familiar with the Sift Verification product to manually test it, as well as have Sift turn the feature on for you. The integration guide can be found here: https://docs.google.com/document/d/11RK_oss2Z7TjydnugYHlG5tK41GMpCyfbNbjzZAYSfM/edit

#### Any questions or concerns about this change?

We have been using this branch at therealreal.com for a few months now so we are quite certain it's performing as expected. There are no changes to any of the existing functionality, just added a bunch of new methods to use for the Verification product. 

#### Deployment Notes

No deployment notes

* [ ] add a reviewer
* [ ] run appropriate tests
* [ ] tag anyone who might want to see this
* [ ] update any relevant wiki documentation

dereksweet committed:
Increase update_account event support

Add support for the following fields for an update_account event
- $session_id
- update_status

dereksweet committed:
Support Verification API

Add support for the Verification API, following 3 endpoints:
- /send
- /resend
- /check

dereksweet committed:
Verification API automated tests

- Basic automated testing added for the Verification product via doctest
- Fixed compiler warnings when running full test suite

dereksweet committed:
Remove incorrect return types

dereksweet committed:
remove unused aliases

dereksweet committed:
Update README

dereksweet committed:
[#173854267] Add support for Verification product

dereksweet committed:
Add support for session_id and verified_event_id

dereksweet committed:
Add support for brand name to verification send payload

dereksweet committed:
allow verified_entity_id to be sent with send events

dereksweet commented on lib/siftsciex/event/account.ex:
```diff
@@ -70,16 +76,16 @@ defmodule Siftsciex.Event.Account do
   ## Examples
 
       iex> Account.create_account(%{user_id: "bob", user_email: "bob@example.com"})
-      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$api_key": api_key, "$type": "$create_account"}
+      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$api_key": api_key(), "$type": "$create_account"}
 
       iex> Account.create_account(%{user_email: "bob@example.com", user_id: "bob", payment_methods: [%{payment_type: :cash}]})
-      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$api_key": api_key, "$payment_methods": [%PaymentMethod{"$payment_type": "$cash"}], "$type": "$create_account"}
+      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$api_key": api_key(), "$payment_methods": [%PaymentMethod{"$payment_type": "$cash"}], "$type": "$create_account"}
 
       iex> Account.create_account(%{user_email: "bob@example.com", user_id: "bob", billing_address: %{name: "Bob", zipcode: "87102"}})
-      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$type": "$create_account", "$api_key": api_key, "$billing_address": %Address{"$name": "Bob", "$zipcode": "87102"}}
+      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$type": "$create_account", "$api_key": api_key(), "$billing_address": %Address{"$name": "Bob", "$zipcode": "87102"}}
 
       iex> Account.create_account(%{user_email: "bob@example.com", user_id: "bob", promotions: [%{promotion_id: "promo"}]})
-      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$type": "$create_account", "$api_key": api_key, "$promotions": [%Promotion{"$promotion_id": "promo"}]}
+      %Account{"$user_email": "bob@example.com", "$user_id": "bob", "$type": "$create_account", "$api_key": api_key(), "$promotions": [%Promotion{"$promotion_id": "promo"}]}
```
This was producing compiler warnings.

dereksweet commented on lib/siftsciex/event/payload/listing.ex:
```diff
@@ -6,7 +6,7 @@ defmodule Siftsciex.Event.Payload.Listing do
 
   alias Siftsciex.Event.Payload
-  alias Siftsciex.Event.Payload.{Address, Image, Item, Location}
+  alias Siftsciex.Event.Payload.{Address, Image, Item}
```
This was producing compiler warnings

dereksweet commented on lib/siftsciex/event/response.ex:
```diff
@@ -25,8 +25,8 @@ defmodule Siftsciex.Event.Response do
 
-      iex> Response.process(%{"status" => 0, "error_message" => "OK", "time" => time, "request" => ""})
-      %Response{status: 0, message: "OK", time: date_time, request: "", score_response: :empty}
+      iex> Response.process(%{"status" => 0, "error_message" => "OK", "time" => time(), "request" => ""})
+      %Response{status: 0, message: "OK", time: date_time(), request: "", score_response: :empty}
```
This was producing compiler warnings

dereksweet commented on lib/siftsciex/event/response/error.ex:
```diff
@@ -43,7 +43,7 @@ defmodule Siftsciex.Event.Response.Error do
   ## Examples
 
-      iex> Error.error(response)
+      iex> Error.error(response())
```
This was producing compiler warnings

dereksweet commented on test/support/mock_request.ex:
```diff
@@ -1,13 +1,13 @@
-  def post(_, %{"$user_id": "transport_error"} = data) do
+  def post(_, %{"$user_id": "transport_error"} = _data) do
     {:error, %{reason: "Timeout"}}
   end
-  def post(_, %{"$user_id": "server_error"} = data) do
+  def post(_, %{"$user_id": "server_error"} = _data) do
     {:ok, %{status_code: 500}}
   end
-  def post(_, %{"$user_id": "client_error"} = data) do
+  def post(_, %{"$user_id": "client_error"} = _data) do
```
This was producing compiler warnings

LeakyBucket commented:
👋 Hi @dereksweet this is awesome!  Unfortunately I'm not with Apartment Therapy any longer, although it looks like @andrewb is maybe he can help out with this?  I would love for improvements to make it into the library so everyone can easily use them.

dereksweet commented:
Sure, thanks for the quick response @LeakyBucket. @andrewb any chance you'd want to merge this? We'd like to get off our local fork to ensure we are up to date with any changes in the future. I'm sure we will be adding more features in the future, as well. Let me know!

andrewb commented:
@dereksweet thanks for your contribution. Unfortunately our team no longer develops our Elixir codebases so I don't expect any updates to this repo in future.

dereksweet commented:
> @dereksweet thanks for your contribution. Unfortunately our team no longer develops our Elixir codebases so I don't expect any updates to this repo in future.

OK that's fair. How would you feel about passing the torch on to me? I would take it up on my personal account.

msramos committed:
Add support for `$login` event

msramos committed:
Bump dependencies

dereksweet commented:
@andrewb I can actually get this merged directly into the SiftSciences organization here: https://github.com/SiftScience

Would you be willing to assist in getting the hexdocs updated to point there instead of at apartmenttherapy?

coryodaniel committed:
Adding a union type for Response (#2)

`__MODULE__.t` on line 27 is cause compilation failures. Adding a union type of the two responses.

dereksweet committed:
Merge pull request #4 from msramos/mr/session-events

Add support for $login event

dereksweet committed:
Update version number and maintainer links

dereksweet committed:
All the code to support getting decisions for an entity

dereksweet committed:
All the supporting tests and mocks

dereksweet committed:
woops use the result type

dereksweet committed:
remove errant commented code

dereksweet committed:
rename decisions_url to accounts_url because it's more accurate

dereksweet commented:
@andrewb thought I'd try one more time to ask you about this. Would you be willing to transfer ownership of this project to The RealReal. You can see we've added more support for both login events and now we are adding support for the Decisions API as well. If not we can just keep adding to our fork or make a new package entirely, but if you'd be willing to pass on ownership of this we have a great relationship with the people at Sift and they will add it to their official packages list. 

Here is some documentation on how to use the hex.owner mix to transfer ownership: https://hexdocs.pm/hex/Mix.Tasks.Hex.Owner.html

dereksweet committed:
add support for sessions and content entities as well

dereksweet committed:
Merge pull request #5 from TheRealReal/story/FUJI-68/add_support_for_decisions_api

[FUJI-68] Add support for Decisions API

dereksweet committed:
Update the URL used for getting a user's score as the old one has been deprecated

dereksweet committed:
update test config as well

dereksweet committed:
Merge pull request #6 from TheRealReal/story/FUJI-92/update_scores_url

[FUJI-92] Update the URL used for getting a user's score

msramos committed:
Sync workflows support

dereksweet committed:
Merge pull request #7 from TheRealReal/mr/add-support-for-sync-workflows-178109577

[#178109577] Synchronous workflow response support

msramos committed:
Better error handling for events

msramos committed:
Add tests for the new event error module

msramos committed:
Update result specs for events

msramos committed:
Fix typo

msramos committed:
Merge pull request #8 from TheRealReal/mr/better-error-handling-for-events

[ZION-661] Better error handling for events

msramos committed:
Fix runtime config for events api

msramos committed:
Merge pull request #9 from TheRealReal/mr/fix-runtime-config

[ZION-691] Evaluate config at runtime
